File: ruby/find-largest-rectangle-in-histogram.rb
#!/usr/bin/env ruby

$arr = [2, 3, 2, 0, 3, 5, 1, 1] # result: 6
# $arr = [2, 3, 2, 0, 3, 5, 1, 9] # result: 9
# $arr = [2, 0, 2, 0, 3, 0, 3, 2] # result: 4

$index = 1;
$reset_index = 0;
$largest_area = 0;

$arr.each_with_index do |v, i|
  i += 1

  if v == 0
    $reset_index = i;
  end

  $index = i - $reset_index

  # get all possible areas at this position
  possible_areas = [v];
  has_secondary_area = $index > 1;

  if has_secondary_area
    # multiply min running value since resetIndex with index
    possible_areas << $arr[$reset_index...i].min * $index
  end

  $largest_area = has_secondary_area ?
    (possible_areas + [$largest_area]).max :
    possible_areas[0]
end

puts $largest_area

File: node/find-largest-rectangle-in-histogram.js
#!/usr/bin/node

const arr = [2, 3, 2, 0, 3, 5, 1, 1]; // result: 6
// const arr = [2, 3, 2, 0, 3, 5, 1, 9]; // result: 9
// const arr = [2, 0, 2, 0, 3, 0, 3, 2]; // result: 4

let index = 1;
let resetIndex = 0;
let largestArea = 0;

arr.forEach((v, i) => {
	i++;

	if( !v ) {
		resetIndex = i;
	}

	index = i - resetIndex;

	// get all possible areas at this position
	let possibleAreas = [v];
	let hasSecondaryArea = index > 1

	if( hasSecondaryArea ) {
		possibleAreas.push(
			// multiply min running value since resetIndex with index
			Math.min.apply(this, (arr.slice(resetIndex, i))) * index
		);
	}

	largestArea = hasSecondaryArea ?
		Math.max(Math.max.apply(this, possibleAreas), largestArea) :
		possibleAreas[0];
});

console.log(largestArea);
